app/models/ems_refresh/parsers/azure.rb

```diff
@@ -0,0 +1,215 @@
+      include Vmdb::Logging
+
+      VALID_LOCATION = /\w+/
+
+      def self.ems_inv_to_hashes(ems, options = nil)
+        new(ems, options).ems_inv_to_hashes
+      end
+
+      def initialize(ems, options = nil)
+        @ems        = ems
+        config      = ems.connect
+        @vmm        = ::Azure::Armrest::VirtualMachineService.new(config)
+        @asm        = ::Azure::Armrest::AvailabilitySetService.new(config)
+        @options    = options || {}
+        @data       = {}
+        @data_index = {}
+      end
+
+      def ems_inv_to_hashes
+        log_header = "MIQ(#{self.class.name}.#{__method__}) Collecting data for EMS : [#{@ems.name}] id: [#{@ems.id}]"
+
+        _log.info("#{log_header}...")
+        get_series
+        get_availability_sets
+        get_instances
+        clean_up_extra_flavor_keys
+        _log.info("#{log_header}...Complete")
+
+        @data
+      end
+
+      private
+
+      def get_series
+        series = []
+        get_locations.each do |location|
+          begin
+            series << @vmm.series(location)
+          rescue RestClient::BadGateway, RestClient::GatewayTimeout, RestClient::BadRequest
+            next
+          end
+        end
+        series = series.flatten
+        series = series.uniq
+        process_collection(series, :flavors) { |s| parse_series(s) }
+      end
+
+      def get_availability_sets
+        a_zones = @asm.list
+        process_collection(a_zones, :availability_zones) { |az| parse_az(az) }
+      end
+
+      def get_instances
+        instances = @vmm.get_vms
+        process_collection(instances, :vms) { |instance| parse_instance(instance) }
+      end
+
+      def process_collection(collection, key)
+        @data[key] ||= []
+
+        return if collection.nil?
+
+        collection.each do |item|
+          uid, new_result = yield(item)
+          @data[key] << new_result
+          @data_index.store_path(key, uid, new_result)
+        end
+      end
+
+      def parse_series(s)
+        name = uid = s['name']
+        new_result = {
+          :type           => "FlavorAzure",
+          :ems_ref        => uid,
+          :name           => name,
+          :cpus           => s['numberOfCores'], # where are the virtual CPUs??
+          :cpu_cores      => s['numberOfCores'],
+          :memory         => s['memoryInMB'].to_f,
+
+          # Extra keys
+          :os_disk_size   => s['osDiskSizeInMB'] * 1024,
+          :swap_disk_size => s['resourceDiskSizeInMB']
+
+        }
+        return uid, new_result
+      end
+
+      def parse_az(az)
+        id = az["id"].downcase
+
+        new_result = {
+          :type    => "AvailabilityZoneAzure",
+          :ems_ref => id,
+          :name    => az["name"],
+        }
+        return id, new_result
+      end
+
+      def parse_instance(instance)
+        uid               = instance.fetch_path('resourceGroup') + "\\" + instance.fetch_path('name')
```

Comment:
The subscription ID is in the connection configuration object. Actually it was not given in configuration step. The first discovered subscription ID was inserted to the configuration object. I think we should allow the user to type subscription ID when adding the azure provider. It can be optional. If not given, the first discovered one can be used.